devlib/utils/ssh.py

```diff
@@ -299,6 +313,498 @@ def __init__(self,
+        logger.info('Setting busybox wrapper to {}'.format(wrapper))
+        self.busybox_wrapper = wrapper
+
+    def set_hostinteractdir(self, indir):
+        logger.info('Setting hostinteractdir  from {} to {}'
+                    .format(self.gem5_input_dir, indir))
+        self.gem5_input_dir = indir
+
+    def connected_as_root(self):
+        # gem5 is always rooted
+        return True
+
+    def push(self, source, dest, timeout=None, wrap_busybox=False):
+        """
+        Push a file to the gem5 device using VirtIO
+
+        The file to push to the device is copied to the temporary directory on
+        the host, before being copied within the simulation to the destination.
+        Checks, in the form of 'ls' with error code checking, are performed to
+        ensure that the file is copied to the destination.
+        """
+        filename = os.path.basename(source)
+        logger.debug("Pushing {} to device.".format(source))
+        logger.debug("gem5interactdir: {}".format(self.gem5_interact_dir))
+        logger.debug("dest: {}".format(dest))
+        logger.debug("filename: {}".format(filename))
+
+        # We need to copy the file to copy to the temporary directory
+        self._move_to_temp_dir(source)
+
+        # Dest in gem5 world is a file rather than directory
+        if os.path.basename(dest) != filename:
+            dest = os.path.join(dest, filename)
+        # Back to the gem5 world
+        self._gem5_shell("ls -al {}{}".format(self.gem5_input_dir,filename))
+        if wrap_busybox:
+            self._gem5_shell("{} cp {}{} {}".format(self.busybox_wrapper,
+                                                    self.gem5_input_dir,
+                                                    filename, dest))
+        else:
+            self._gem5_shell("cat {}{} > {}".format(self.gem5_input_dir, filename,
+                                                    dest))
+        self._gem5_shell("sync")
+        self._gem5_shell("ls -al {}".format(dest))
+        self._gem5_shell("ls -al {}".format(self.gem5_input_dir))
+        logger.debug("Push complete.")
+
+    def pull(self, source, dest, timeout=0, wrap_busybox=False): #pylint: disable=unused-argument
+        """
+        Pull a file from the gem5 device using m5 writefile
+
+        The file is copied to the local directory within the guest as the m5
+        writefile command assumes that the file is local. The file is then
+        written out to the host system using writefile, prior to being moved to
+        the destination on the host.
+        """
+        filename = os.path.basename(source)
+
+        logger.debug("pull_file {} {}".format(source, filename))
+        # We don't check the exit code here because it is non-zero if the source
+        # and destination are the same. The ls below will cause an error if the
+        # file was not where we expected it to be.
+        self._gem5_shell("{} cp {} {}".format(self.busybox_wrapper, source,
```

Comment:
why is busybox needed here?

Also, the command should be `"cp '{}' '{}'"`. Source and/or destination paths may have spaces, parentheses, or other shell characters in their names.